### ISSUE TITLE
CacheConcurrencyStrategy.READ_WRITE not working

### ISSUE THREAD

**Issue by antoniosch85** [jhipster/generator-jhipster#17425]
##### **Overview of the issue**

The annotation CacheConcurrencyStrategy.READ_WRITE seems not working. Always an empty result is returned, I need to restart server to see the correct return.

##### **Motivation for or Use Case**

I am not able to use the cache. In order to work, I have to eliminate the annotation 
@Cache(usage = CacheConcurrencyStrategy.READ_WRITE)

##### **Reproduce the error**

I have 2 entities A and B with one to many relationship.
I generated all service and in the entity A I can see:

    @OneToMany(mappedBy = "entityA")
    @Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
    private Set<entityB> entitiesB = new HashSet<>();

If I add a new entity B and set the relationship with entity A, my expectation is when I call entityA.getEntitiesB() the inserted entityB is returned but I get always an empty set.
If I restart the server the correct list is returned.

##### **JHipster Version(s)**

v6.10.5

##### **Browsers and Operating System**

- [X] Checking this box is mandatory (this is just to show you read everything)

**Comment by pmverma** [comment]
I notice the same issue with my application and today I can confirm it.
This is really a serious bug since we are getting incorrect results with OneToMany.

Here is a reproducible jdl.

```
application {
  config {
    applicationType monolith
    authenticationType jwt
    baseName jhcache
    blueprints []
    buildTool maven
    cacheProvider hazelcast
    clientFramework angularX
    clientPackageManager npm
    clientTheme none
    creationTimestamp 1640705957630
    databaseType sql
    devDatabaseType h2Disk
    dtoSuffix DTO
    enableGradleEnterprise false
    enableHibernateCache true
    enableSwaggerCodegen false
    enableTranslation false
    gradleEnterpriseHost ""
    jhiPrefix jhi
    jhipsterVersion "7.4.1"
    jwtSecretKey "ZWJmMTkzOTE3MjY2OTYwYmIwZGQxMTk1YWJiYmU2N2FkMTI1Y2NlMDY3MGI2Y2YyOWZmY2U1MTA2MDc2Mjk5OTNkYmU2NjU5N2EwOGVhOWE4ZWJmNTAyMDRkNTk4MWI1N2U2ZjYxYTMzMmM4ZmI3M2JkOWJhMzc3MjJmYjAxYjY="
    languages [en, fr]
    messageBroker false
    nativeLanguage en
    otherModules []
    packageName com.mycompany.myapp
    prodDatabaseType postgresql
    reactive false
    searchEngine false
    serverPort 8080
    serviceDiscoveryType no
    skipUserManagement false
    testFrameworks []
    websocket false
    withAdminUi true
  }

  entities Child, Parent
}

entity Child {
  name String
}
entity Parent {
  name String
}
relationship OneToMany {
  Parent{child} to Child{parent(name)}
}

dto Child, Parent with mapstruct
service Child, Parent with serviceClass

``` 

Steps to reproduce

1. Generate application with above jdl file. `jhipster import-jdl jdlcontent.jh`
2. Add the following code to `ParentService.findOne` method to print `-ToMany` side.
```
parentRepository.findById(id).ifPresent(parent -> parent.getChildren().forEach(System.out::println));
```
3. Start the application
4. Add a parent entity to a child's entity and visit the parent details page.
5. Add the same parent to another child and visit again to the same parent details page.
6. You can see that `ParentService.findOne` method is not printing the correct number of child entities.

I am checking for the same issue however I am wondering how many applications have been affected since it is reported for `v6.10.5`.

**Comment by pascalgrimaud** [comment]
Adding a bounty as the bug is confirmed by you, @pmverma 
If you know how to fix it, plz go ahead!

**Comment by atomfrede** [comment]
It looks like the problem seems to be related to the caching of the old parent entity. When assigning a new parent the old one is not changed and stays in the cache. When you save/update the old parent the cache is consistent again. Not sure if thats the root cause and if thats how the hibernate cache is supposed to work.

**Comment by antoniosch85** [comment]
> It looks like the problem seems to be related to the caching of the old parent entity. When assigning a new parent the old one is not changed and stays in the cache. When you save/update the old parent the cache is consistent again. Not sure if thats the root cause and if thats how the hibernate cache is supposed to work.

If you update the parent the cache is updated, the issue occurs only when a new parent has been inserted.
I think it is how the hibernate cache is supposed to work.

**Comment by pmverma** [comment]
The issue here is we are not synchronizing bi-directional associations although we already have the synchronization code in parent classes.
[How to synchronize bidirectional entity associations with JPA and Hibernate](https://vladmihalcea.com/jpa-hibernate-synchronize-bidirectional-entity-associations/)
Naturally, we would manage the synchronization from the parent side however if you want to do in child class/service it should work but then again at least we would need to load the parent so we can add a child to it.

Two solutions I can think of as of now.

```
Code snippet of ChildService.java
/**
 * Save a child.
 *
 * @param childDTO the entity to save.
 * @return the persisted entity.
 */
public ChildDTO save(ChildDTO childDTO) {
    log.debug("Request to save Child : {}", childDTO);
    Child child = childMapper.toEntity(childDTO);
    // solution1, syncing starts here
    if (Objects.nonNull(child.getParent()) && Objects.nonNull(child.getParent().getId())) {
        final var parentOptional = parentRepository.findById(child.getParent().getId());
        if (parentOptional.isPresent()) {
            final var parent = parentOptional.get();
            parent.addChild(child);
        }
    }
    // solution1, syncing ends here

    child = childRepository.save(child);

    // solution2, syncing starts here
    child.getParent().addChild(child);
    // solution2, syncing ends here

    return childMapper.toDto(child);
}

```
Solution1 is a more traditional approach but has some concerns.

- More code just for syncing. Assume a child entity has multiple parent entities ( for eg, one person has multiple people to report to). In that case, the above-syncing code will be needed for each person.
- Dependency on each parent `Service/Repository` class for loading the parent entity. JHipster does not generate such code as of today.

Solution2 seems weird but it works and we can use the code we already have. No dependency. Not many lines of code even we have multiple parents.

I would rather like to wait for what experts say.
cc @jhipster/developers

**Comment by atomfrede** [comment]
I totally agree, loading the parent inside child service is ugly (tried it too). So maybe we can somehow decouple it with some spring events (or jpa events?).

**Comment by Tcharl** [comment]
V8 could be the occasion to go modulith

**Comment by pmverma** [comment]
For v7, if there is no major concern I would like to go with solution2.
For v8 or future major versions, we still can have some other suitable solution.

**Comment by DanielFran** [comment]
@pmverma Are you still available to do a PR?

**Comment by ReginaBDR** [comment]
@OneToMany(cascade = { CascadeType.MERGE, CascadeType.PERSIST }) does the trick

**Comment by Asobu01** [comment]
I can take jhipster/generator-jhipster#17425.
For "JHipster is not synchronizing bidirectional entity associations cache", I will start by reproducing the server-side issue and isolating root cause with logs/tests.
Target: draft PR in ~24h.
Deliverables: repro note with concrete evidence (logs/trace when relevant); focused backend patch + tests/log evidence; PR with verification steps.
@Asobu01

**Comment by Asobu01** [comment]
I can take jhipster/generator-jhipster#17425.
For "JHipster is not synchronizing bidirectional entity associations cache", I will start by reproducing the server-side issue and isolating root cause with logs/tests.
Target: draft PR in ~24h.
Deliverables: repro note with concrete evidence (logs/trace when relevant); focused backend patch + tests/log evidence; PR with verification steps.
@Asobu01

**Comment by Asobu01** [comment]
I can take jhipster/generator-jhipster#17425.
For "JHipster is not synchronizing bidirectional entity associations cache", I will start by reproducing the server-side issue and isolating root cause with logs/tests.
Target: draft PR in ~24h.
Deliverables: repro note with concrete evidence (logs/trace when relevant); focused backend patch + tests/log evidence; PR with verification steps.
@Asobu01

**Comment by qmonmert** [comment]
@mshima We're in an infinite loop here (comments) 😅

cc @DanielFran @jdubois @mraible

**Comment by avdev4j** [comment]
@qmonmert I suspect we are facing AI bots here. We have same issues on other threads. Do they target bounties tickets ?

**Comment by qmonmert** [comment]
@avdev4j yes this is why @mshima created this ticket https://github.com/jhipster/generator-jhipster/issues/32604

**Comment by avdev4j** [comment]
Yep, I saw it just after my comment 😁

**Comment by jdubois** [comment]
Hey, are you facing some bot spamming here? If that's the case, I can lock discussions or block bots...

**Comment by atomfrede** [comment]
Please Block the recent Bots imho. Also see https://github.com/jhipster/generator-jhipster/issues/32604

**Comment by jdubois** [comment]
Conversation locked, and user blocked.
